Lib9c/Model/Order/Order.cs

```diff
@@ -83,7 +88,55 @@ public virtual void Validate(AvatarState avatarState, int count)
             }
         }
 
-        public virtual void ValidateCancelOrder(AvatarState avatarState, Guid tradableId)
+        public ITradableItem Cancel(AvatarState avatarState, long blockIndex)
+        {
+            if (avatarState.inventory.TryGetLockedItem(new OrderLock(OrderId), out Inventory.Item inventoryItem))
+            {
+                inventoryItem.Unlock();
+                var tradableItem = (ITradableItem)inventoryItem.item;
+                tradableItem.RequiredBlockIndex = blockIndex;
+                return tradableItem;
+            }
+            throw new ItemDoesNotExistException(
+                $"Aborted because the tradable item({TradableId}) was failed to load from avatar's inventory.");
+        }
+
+        public void ValidateCancelOrder(AvatarState avatarState, Guid tradableId)
+        {
+            if (!avatarState.address.Equals(SellerAvatarAddress) || !avatarState.agentAddress.Equals(SellerAgentAddress))
+            {
+                throw new InvalidAddressException($"Invalid Seller Addresses. Expected Addresses: {SellerAgentAddress}, {SellerAvatarAddress}. Actual: {avatarState.agentAddress}, {avatarState.address}");
+            }
+
+            if (!TradableId.Equals(tradableId))
+            {
+                throw new InvalidTradableIdException($"{tradableId} is not equals {TradableId}");
+            }
+
+            if (!avatarState.inventory.TryGetLockedItem(new OrderLock(OrderId), out Inventory.Item inventoryItem))
+            {
+                throw new ItemDoesNotExistException(
+                    $"Aborted because the tradable item({TradableId}) was failed to load from avatar's inventory.");
+            }
+
+            var itemCount = this is FungibleOrder fungibleOrder ? fungibleOrder.ItemCount : 1;
```

Comment:
IMO, it's more natural using polymorphism (implmenting `FungibleOrder.ItemCount` / `NonFungibleOrder.ItemCount` separately) instead of type query(`is FungibleOrder`)